File: endpoints/python/test_mutable.py
#!/usr/bin/env python
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 ZeroDDS Contributors
#
# Byte-identity test for the Python EMHEADER / @mutable path (ADR 0013).
# Same vector as endpoints/golden-gen encode_mutable.
#
# usage: python test_mutable.py <golden_mutable_le.bin> <golden_mutable_be.bin>

from __future__ import print_function

import sys

import zerodds_wire as zw


def encode(endian):
    w = zw.Writer(endian)
    bs = w.dheader_begin()                 # @mutable struct DHEADER
    e = w.emheader_begin(10, False); w.put_u32(0xDEADBEEF); w.emheader_end(e)
    e = w.emheader_begin(20, False); w.put_string(u"mut"); w.emheader_end(e)
    e = w.emheader_begin(30, False); w.put_u16(0x0777); w.emheader_end(e)
    w.dheader_end(bs)
    return w.bytes()


def decode(data, endian):
    r = zw.Reader(data, endian)
    dh = r.dheader_read()
    start = r.pos
    out = {}
    while r.pos - start < dh:
        member_id, _mu, nextint = r.emheader_read()
        if member_id == 10:
            out["x"] = r.get_u32()
        elif member_id == 20:
            out["s"] = r.get_string()
        elif member_id == 30:
            out["k"] = r.get_u16()
        else:
            r.get_bytes(nextint)  # skip unknown
    return out


def check(endian, path, tag):
    f = open(path, "rb")
    golden = f.read()
    f.close()
    out = encode(endian)
    if out != golden:
        print("%s: MISMATCH len py=%d golden=%d" % (tag, len(out), len(golden)))
        return 1
    print("%s: %d bytes byte-identical to Rust golden (EMHEADER/@mutable)"
          % (tag, len(out)))
    d = decode(out, endian)
    if d.get("x") != 0xDEADBEEF or d.get("s") != u"mut" or d.get("k") != 0x0777:
        print("%s: round-trip mismatch %r" % (tag, d))
        return 1
    print("%s: round-trip decode ok" % tag)
    return 0


def main():
    if len(sys.argv) < 3:
        print("usage: %s <mutable_le> <mutable_be>" % sys.argv[0])
        return 2
    rc = check(zw.LE, sys.argv[1], "LE") | check(zw.BE, sys.argv[2], "BE")
    if rc == 0:
        print("ALL OK")
    return rc


if __name__ == "__main__":
    sys.exit(main())

File: endpoints/cpp/test/test_mutable.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ZeroDDS Contributors
//
// Byte-identity test for the C++98 EMHEADER / @mutable path (ADR 0013), via
// the facade over the C wire-core. Same vector as endpoints/golden-gen.
//
// usage: test_mutable <golden_mutable_le.bin> <golden_mutable_be.bin>

#include <cstdio>
#include <cstring>

#include "zerodds_wire.hpp"

static long read_file(const char* path, unsigned char* buf, long cap) {
    std::FILE* f = std::fopen(path, "rb");
    if (f == 0) { std::fprintf(stderr, "cannot open %s\n", path); return -1; }
    long n = (long)std::fread(buf, 1, (size_t)cap, f);
    std::fclose(f);
    return n;
}

static int check(int endian, const char* golden_path, const char* tag) {
    unsigned char out[256], golden[256];
    zerodds::Writer w(out, sizeof(out), endian);
    size_t bs = w.dheader_begin();
    size_t e;
    e = w.emheader_begin(10, false); w.u32(0xDEADBEEFuL);  w.emheader_end(e);
    e = w.emheader_begin(20, false); w.str("mut");          w.emheader_end(e);
    e = w.emheader_begin(30, false); w.u16(0x0777u);        w.emheader_end(e);
    w.dheader_end(bs);
    if (w.error() != ZDW_OK) { std::fprintf(stderr, "%s: encode err\n", tag); return 1; }

    long gn = read_file(golden_path, golden, (long)sizeof(golden));
    if (gn < 0) return 1;
    if ((long)w.size() != gn || std::memcmp(out, golden, (size_t)gn) != 0) {
        std::fprintf(stderr, "%s: byte mismatch (C++=%ld golden=%ld)\n",
                     tag, (long)w.size(), gn);
        return 1;
    }
    std::printf("%s: %ld bytes byte-identical to Rust golden (EMHEADER/@mutable)\n",
                tag, (long)w.size());

    zerodds::Reader r(out, w.size(), endian);
    unsigned long dh = r.dheader();
    unsigned long x = 0, k = 0, ni = 0;
    std::string s;
    for (int i = 0; i < 3; i++) {
        unsigned long id = r.emheader(&ni);
        if (id == 10) x = r.u32();
        else if (id == 20) s = r.str();
        else if (id == 30) k = r.u16();
    }
    if (r.error() != ZDW_OK || dh == 0 || x != 0xDEADBEEFuL || s != "mut" || k != 0x0777u) {
        std::fprintf(stderr, "%s: round-trip mismatch\n", tag);
        return 1;
    }
    std::printf("%s: round-trip decode ok\n", tag);
    return 0;
}

int main(int argc, char** argv) {
    if (argc < 3) { std::fprintf(stderr, "usage: %s le be\n", argv[0]); return 2; }
    int rc = check(ZDW_LE, argv[1], "LE") | check(ZDW_BE, argv[2], "BE");
    if (rc == 0) std::printf("ALL OK\n");
    return rc;
}
